File: admin/assets/examples/licora-v2/python/licora_v2_client.py
#!/usr/bin/env python3
"""Licora Secure API v2 lifecycle reference.

This developer/test client creates an ephemeral device credential, exercises
activate -> status -> refresh -> status -> deactivate, then exits. Production
applications must persist the device private key and rotated refresh token in
OS-backed secure storage and verify LICORA-V2/RS256 access-token signatures
with the pinned Licora server public key before trusting token claims locally.
"""
from __future__ import annotations

import argparse
import base64
import hashlib
import json
import secrets
import time
import uuid
from urllib.parse import urljoin, urlparse

import requests
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec


def b64url(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def jwt_payload(token: str) -> dict:
    parts = token.split(".")
    if len(parts) != 3:
        raise RuntimeError("Licora returned a malformed access token")
    raw = parts[1] + "=" * ((4 - len(parts[1]) % 4) % 4)
    return json.loads(base64.urlsafe_b64decode(raw.encode("ascii")))


def canonical(method: str, path: str, timestamp: int, nonce: str, body: bytes, context: str) -> bytes:
    return "\n".join([
        method.upper(), path, str(timestamp), nonce,
        hashlib.sha256(body).hexdigest().lower(), context,
    ]).encode("utf-8")


def compact_json(value: dict) -> bytes:
    return json.dumps(value, separators=(",", ":"), ensure_ascii=False).encode("utf-8")


class LicoraV2:
    def __init__(self, base_url: str, app_id: str, app_version: str) -> None:
        self.base_url = base_url.rstrip("/") + "/"
        self.app_id = app_id
        self.app_version = app_version
        self.private_key = ec.generate_private_key(ec.SECP256R1())
        self.public_pem = self.private_key.public_key().public_bytes(
            serialization.Encoding.PEM,
            serialization.PublicFormat.SubjectPublicKeyInfo,
        ).decode("ascii")
        self.device_id = "py-" + uuid.uuid4().hex

    def endpoint(self, name: str) -> str:
        return urljoin(self.base_url, f"api/v2/{name}.php")

    def proof_headers(self, url: str, body: bytes, context: str, access_token: str | None = None) -> dict[str, str]:
        timestamp = int(time.time())
        nonce = b64url(secrets.token_bytes(18))
        path = urlparse(url).path or "/"
        signature = self.private_key.sign(
            canonical("POST", path, timestamp, nonce, body, context),
            ec.ECDSA(hashes.SHA256()),
        )
        headers = {
            "Content-Type": "application/json",
            "X-Licora-Timestamp": str(timestamp),
            "X-Licora-Nonce": nonce,
            "X-Licora-Device-Signature": b64url(signature),
        }
        if access_token:
            headers["Authorization"] = "Bearer " + access_token
        return headers

    def post(self, name: str, payload: dict, context: str, access_token: str | None = None) -> dict:
        url = self.endpoint(name)
        body = compact_json(payload)
        response = requests.post(
            url,
            data=body,
            headers=self.proof_headers(url, body, context, access_token),
            timeout=20,
        )
        try:
            data = response.json()
        except ValueError as exc:
            raise RuntimeError(f"HTTP {response.status_code}: non-JSON response") from exc
        if not data.get("success"):
            raise RuntimeError(f"Licora error {data.get('code', 'UNKNOWN')} (HTTP {response.status_code})")
        return data

    def activate(self, license_key: str) -> dict:
        payload = {
            "license_key": license_key,
            "app_id": self.app_id,
            "app_version": self.app_version,
            "device_id": self.device_id,
            "device_public_key": self.public_pem,
        }
        return self.post("activate", payload, "activate:" + self.app_id)

    def status(self, access_token: str) -> dict:
        jti = str(jwt_payload(access_token)["jti"])
        return self.post("status", {}, jti, access_token)

    def refresh(self, refresh_token: str) -> dict:
        payload = {"refresh_token": refresh_token, "app_version": self.app_version}
        context = "refresh:" + hashlib.sha256(refresh_token.encode("utf-8")).hexdigest()
        return self.post("refresh", payload, context)

    def deactivate(self, access_token: str) -> dict:
        jti = str(jwt_payload(access_token)["jti"])
        return self.post("deactivate", {}, jti, access_token)


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--base-url", required=True)
    parser.add_argument("--app-id", required=True)
    parser.add_argument("--license-key", required=True)
    parser.add_argument("--app-version", default="1.0.0")
    args = parser.parse_args()

    client = LicoraV2(args.base_url, args.app_id, args.app_version)
    access_token = ""
    try:
        activated = client.activate(args.license_key)
        access_token = str(activated["access_token"])
        refresh_token = str(activated["refresh_token"])
        print("[PASS] activate", activated["code"])

        current = client.status(access_token)
        print("[PASS] status", current["code"], current.get("license", {}).get("status"))

        refreshed = client.refresh(refresh_token)
        access_token = str(refreshed["access_token"])
        refresh_token = str(refreshed["refresh_token"])
        print("[PASS] refresh", refreshed["code"], "rotated refresh token")

        current = client.status(access_token)
        print("[PASS] status-after-refresh", current["code"])

        client.deactivate(access_token)
        access_token = ""
        print("[PASS] deactivate")
        return 0
    finally:
        if access_token:
            try:
                client.deactivate(access_token)
                print("[INFO] cleanup deactivate completed")
            except Exception as exc:  # cleanup only
                print("[WARN] cleanup deactivate failed:", exc)


if __name__ == "__main__":
    raise SystemExit(main())

File: admin/assets/examples/licora-v2/node/licora-v2-client.mjs
#!/usr/bin/env node
/**
 * Licora Secure API v2 lifecycle reference.
 *
 * Ephemeral developer/test flow: activate -> status -> refresh -> status ->
 * deactivate. Production apps must persist the P-256 private key and rotated
 * refresh token securely and verify LICORA-V2/RS256 tokens with the pinned
 * server public key before trusting token claims locally.
 */
import {
  createHash,
  createSign,
  generateKeyPairSync,
  randomBytes,
} from 'node:crypto';

const b64url = (buffer) => Buffer.from(buffer).toString('base64').replace(/=/g, '').replace(/\+/g, '-').replace(/\//g, '_');
const sha256 = (value) => createHash('sha256').update(value).digest('hex');
const compact = (value) => Buffer.from(JSON.stringify(value), 'utf8');

function decodeJwtPayload(token) {
  const parts = token.split('.');
  if (parts.length !== 3) throw new Error('Licora returned a malformed access token');
  return JSON.parse(Buffer.from(parts[1], 'base64url').toString('utf8'));
}

class LicoraV2Client {
  constructor(baseUrl, appId, appVersion) {
    this.baseUrl = baseUrl.replace(/\/+$/, '');
    this.appId = appId;
    this.appVersion = appVersion;
    const keys = generateKeyPairSync('ec', {
      namedCurve: 'prime256v1',
      publicKeyEncoding: { type: 'spki', format: 'pem' },
      privateKeyEncoding: { type: 'pkcs8', format: 'pem' },
    });
    this.publicKey = keys.publicKey;
    this.privateKey = keys.privateKey;
    this.deviceId = `node-${randomBytes(16).toString('hex')}`;
  }

  endpoint(name) { return `${this.baseUrl}/api/v2/${name}.php`; }

  headers(url, body, context, accessToken = '') {
    const timestamp = Math.floor(Date.now() / 1000);
    const nonce = b64url(randomBytes(18));
    const path = new URL(url).pathname || '/';
    const canonical = ['POST', path, timestamp, nonce, sha256(body), context].join('\n');
    const signer = createSign('SHA256');
    signer.update(canonical, 'utf8');
    signer.end();
    const signature = signer.sign({ key: this.privateKey, dsaEncoding: 'der' });
    return {
      'Content-Type': 'application/json',
      'X-Licora-Timestamp': String(timestamp),
      'X-Licora-Nonce': nonce,
      'X-Licora-Device-Signature': b64url(signature),
      ...(accessToken ? { Authorization: `Bearer ${accessToken}` } : {}),
    };
  }

  async post(name, payload, context, accessToken = '') {
    const url = this.endpoint(name);
    const body = compact(payload);
    const response = await fetch(url, { method: 'POST', headers: this.headers(url, body, context, accessToken), body });
    const data = await response.json().catch(() => { throw new Error(`HTTP ${response.status}: non-JSON response`); });
    if (!data.success) throw new Error(`Licora error ${data.code || 'UNKNOWN'} (HTTP ${response.status})`);
    return data;
  }

  activate(licenseKey) {
    return this.post('activate', {
      license_key: licenseKey,
      app_id: this.appId,
      app_version: this.appVersion,
      device_id: this.deviceId,
      device_public_key: this.publicKey,
    }, `activate:${this.appId}`);
  }

  status(accessToken) {
    return this.post('status', {}, String(decodeJwtPayload(accessToken).jti), accessToken);
  }

  refresh(refreshToken) {
    return this.post('refresh', { refresh_token: refreshToken, app_version: this.appVersion }, `refresh:${sha256(refreshToken)}`);
  }

  deactivate(accessToken) {
    return this.post('deactivate', {}, String(decodeJwtPayload(accessToken).jti), accessToken);
  }
}

async function main() {
  const [baseUrl, appId, licenseKey, appVersion = '1.0.0'] = process.argv.slice(2);
  if (!baseUrl || !appId || !licenseKey) {
    console.error('Usage: node licora-v2-client.mjs <base-url> <app-id> <license-key> [app-version]');
    process.exitCode = 2;
    return;
  }
  const client = new LicoraV2Client(baseUrl, appId, appVersion);
  let accessToken = '';
  try {
    const activated = await client.activate(licenseKey);
    accessToken = activated.access_token;
    let refreshToken = activated.refresh_token;
    console.log('[PASS] activate', activated.code);
    console.log('[PASS] status', (await client.status(accessToken)).code);
    const refreshed = await client.refresh(refreshToken);
    accessToken = refreshed.access_token;
    refreshToken = refreshed.refresh_token;
    console.log('[PASS] refresh', refreshed.code, 'rotated refresh token');
    console.log('[PASS] status-after-refresh', (await client.status(accessToken)).code);
    await client.deactivate(accessToken);
    accessToken = '';
    console.log('[PASS] deactivate');
  } finally {
    if (accessToken) {
      try { await client.deactivate(accessToken); console.log('[INFO] cleanup deactivate completed'); }
      catch (error) { console.warn('[WARN] cleanup deactivate failed:', error.message); }
    }
  }
}

if (import.meta.url === `file://${process.argv[1]}`) main().catch((error) => { console.error(error.message); process.exitCode = 1; });

export { LicoraV2Client };
